src/com/hms_networks/americas/sc/realtimedata/RealTimeDataQueueManager.java

```diff
@@ -0,0 +1,197 @@
+    }
+  }
+
+  /**
+   * Tag iterators are used to retrieve a single data point from each tag before going to the next
+   * tag in a tag group list. Initialize all tag iterators to 0 to start at the beginning of the
+   * lists.
+   */
+  private static void initTagManageIterators() {
+
+    tagManagerIterators = new int[NUM_TAG_GROUPS];
+
+    for (int i = 0; i < NUM_TAG_GROUPS; i++) {
+      tagManagerIterators[i] = 0;
+    }
+  }
+
+  /** Initializes RealTimeDataQueueManager */
+  public static void init() {
+    initTagGroups();
+    initTagManagers();
+    initTagManageIterators();
+  }
+
+  /**
+   * getGroupANextData will group a single data point object from the tags in a group, rotate
+   * iterators to look at the next tag in the list of that group's tags, and remove the retrieved
+   * tag from the list of tags.
+   *
+   * <p>If there are no DataPoints left, null will be returned.
+   *
+   * @param tagGroup The tag group to select a datapoint from
+   * @return returns a DataPoint object
+   */
+  public static DataPoint getGroupNextData(int tagGroup) {
+    DataPoint data = null;
+
+    RealTimeTagDataPointManager dataPoints =
+        ((RealTimeTagDataPointManager)
+            ((ArrayList) tagManagers.get(tagGroup)).get(tagManagerIterators[tagGroup]));
+    if (dataPoints.getSize() > 0) {
+      data = dataPoints.removeDataPoint();
+      tagManagerIterators[tagGroup] += 1;
+      tagManagerIterators[tagGroup] =
+          (tagManagerIterators[tagGroup] % ((ArrayList) tagGroupList.get(tagGroup)).size());
+    }
+    return data;
+  }
+
+  /**
+   * Gets the number of data points retrieved from tag a group.
+   *
+   * @param tagGroup The tag group to get the number of data points from
+   * @return returns the number of data points in specified group
+   */
+  public static int getNumGroupDataPoints(int tagGroup) {
+    int numDataPoints = 0;
+
+    for (int i = 0; i < ((ArrayList) tagManagers.get(tagGroup)).size(); i++) {
+      numDataPoints +=
+          ((RealTimeTagDataPointManager) ((ArrayList) tagManagers.get(tagGroup)).get(i)).getSize();
+    }
+
+    return numDataPoints;
+  }
+
+  /**
+   * Gets a new data point for each tag in group A and adds it to an internal queue in the tag's tag
+   * Manager.
+   *
+   * @param tagGroup The tag group to fetch data from.
+   */
+  public static void dataFetcher(int tagGroup) {
+    for (int i = 0; i < ((ArrayList) tagManagers.get(tagGroup)).size(); i++) {
+      // tag group and tag manager are the same size
+      TagInfo currentTag = ((TagInfo) ((ArrayList) tagGroupList.get(tagGroup)).get(i));
+      TagControl tagControl =
+          ((RealTimeTagDataPointManager) ((ArrayList) tagManagers.get(tagGroup)).get(i)).tagControl;
+      DataPoint data = getCurentTagValue(currentTag, tagControl);
+      // add new data point to the current tag manager
+      ((RealTimeTagDataPointManager) ((ArrayList) tagManagers.get(tagGroup)).get(i))
+          .addDataPoint(data);

```

Comment:
So this function looks up the tagcontrol from inside the tagmanager, reads the value, then looks up the tagmanager again to put that value in? Why cant tagManager.addDataPoint handle everything on its own?